### ISSUE TITLE
"Passing nesting to interpolated" is a risky default

### ISSUE THREAD

**Issue by lynn** [i18next/i18next#1479]
I am not so sure if [_Passing nesting to interpolated_](https://www.i18next.com/translation-function/nesting#passing-nesting-to-interpolated) should be a feature at all. It invites XSS-like behavior.

If an app has localization strings like:

```json
  "new-message": "You have a new message from {{name}}.",
  "foo": "Lorem ipsum dolor sit amet, ..."
```

Then this feature being active by default means I can adjust my name in the app from `lynn` to `$t(foo)` to cause all sorts of confusing behavior for anyone who is shown that string (here, for example, causing a huge Lorem Ipsum text to display instead of my name).

Or I can set my name in the app to `$t(new-message)` and overflow the call stack:

![image](https://user-images.githubusercontent.com/16232127/86972287-00615500-c173-11ea-8733-493c617dc94b.png)

I _really really_ do not expect i18next to specially parse strings passed as localization parameters like this.

If I am supposed to somehow escape _all_ user-supplied input that's ever passed into localization as arguments, then the documentation should be very clear about this, and fix a lot of examples. But it doesn't seem worth it, because I don't really understand the point of this feature at all. I can just write `i18next.t('key2', {val: i18next.t('key1')});` instead of the example in the docs, right?

I suggest that this feature be turned off by default, and maybe removed entirely. (Nesting itself is fine, but the `$t()` syntax should _not_ be parsed in parameters, only in the localization strings themselves.)

If I am overestimating the riskiness of this feature please let me know.

**Comment by jamuhl** [comment]
> I really really do not expect i18next to specially parse strings passed as localization parameters like this.

...others did -> and provided PR to make this possible...

i18next per default does escapes...there is no way per default to inject a script doing an xss attack...

for interpolation we got some check that it does not result in an endless loop...https://github.com/i18next/i18next/blob/master/src/Interpolator.js#L102 seems the newly added feature does circumvent that.

**Comment by adrai** [comment]
Should be fixed with v19.5.5

**Comment by adrai** [comment]
If you like this module don’t forget to star this repo. Make a tweet, share the word or have a look at our https://locize.com to support the devs of this project.

There are many ways to help this project 🙏

**Comment by lynn** [comment]
@adrai @jamuhl The infinite recursion isn't the core problem I'm addressing here: it's the fact that I, the user of an i18next app, can set a string like “my user name in that app” to something like `$t(foo)` and trigger an interpolation where it _isn't meant to happen at all_.

If I set my name to `$t(foo)` (even if that is a crazy name) it should show up *untouched* as: “**$t(foo)** is typing...”

Instead, it will trigger this "Passing nesting to interpolated" feature and become some long unrelated localization string: “**Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua** is typing…”

Maybe XSS was the wrong word. I know that i18next will escape `<>` tags and prevents actual script execution. The worst that can happen here is really: “arbitrary localized app text ending up on the screen instead of the user's name”. But that can _still_ be pretty nasty: if `foo` maps to a very long string, I can make the app very annoying to use for everyone who has to look at my 1,000+ character interpolated name. Or, if `Administrator` is a disallowed name, I can maybe set my name to `$t(admin)` instead, which becomes `Administrator` on another user's screen, and do phishing.

I simply find this all quite dangerous. How do I stop this behavior from happening? I hope you agree that “oh, just disallow `$` in user names” isn't a real solution.

> others did -> and provided PR to make this possible

That doesn't mean it's out of the question that there may be security concerns surrounding the feature.

**Comment by adrai** [comment]
Maybe I misunderstood, but If you want to stop interpolation of nested $(foo) just try to set `nestingPrefix` and `nestingSuffix` to something that would not match. => https://www.i18next.com/translation-function/interpolation#all-interpolation-options

**Comment by lynn** [comment]
Can I set them to `undefined` to disable the feature entirely? I don't think it's hard for an attacker to retrieve `nestingPrefix` and `nestingSuffix` and abuse this feature to trigger unintended interpolations, no matter what I set them to.

**Comment by adrai** [comment]
The PR #1481 would disable all the usage with `$t()` probably this is not what you are looking for... your cat example would not work anymore

**Comment by adrai** [comment]
btw: https://github.com/i18next/i18next/issues/854

**Comment by adrai** [comment]
@lynn can you have a look at: https://github.com/i18next/i18next/pull/1483 ?

**Comment by lynn** [comment]
@adrai This looks great! Thank you.

I hope you will consider making `true` the default value of `skipOnVariables`. Or at least, prominently suggest in the documentation that it should be set to `true` and that this feature is deprecated.

**Comment by adrai** [comment]
Released with v19.6.0
